File: backend/ticket_service/src/main/java/com/ticketflow/ticket_service/api/TicketsApiController.java
package com.ticketflow.ticket_service.api;

import com.ticketflow.ticket_service.service.TicketsService;

import java.util.List;

import com.ticketflow.ticket_service.models.Ticket;
import com.ticketflow.ticket_service.models.client_models.OrderModel;
import com.ticketflow.ticket_service.models.exceptions.NotFoundException;
import com.ticketflow.ticket_service.models.exceptions.TicketAlreadyOrderedException;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RestController;

@RestController
public class TicketsApiController {

    private TicketsService ticketsService;

    @Autowired
    public TicketsApiController(TicketsService ticketsService) {
        this.ticketsService = ticketsService;
    }

    @GetMapping(value = "/")
    public ResponseEntity<String> home() {
        return ResponseEntity.ok().body("<span>Hello from TicketFlow Ticket service</span>");
    }

    @GetMapping(value = "tickets/by-movie/{id}")
    public ResponseEntity<List<Ticket>> getByMovieId(@PathVariable Integer movieId) {
        return ResponseEntity.ok(ticketsService.getByMovieId(movieId));
    }

    @PostMapping(value = "/tickets/by-user")
    public ResponseEntity<List<Ticket>> getByUserEmail(@RequestBody String userEmail) {
        return ResponseEntity.ok(ticketsService.getByUserEmail(userEmail));
    }

    @PostMapping(value = "/tickets/order")
    public ResponseEntity<String> order(@RequestBody OrderModel order)
            throws TicketAlreadyOrderedException, NotFoundException {
        ticketsService.order(order);
        return ResponseEntity.accepted().body("Ordered successfully");
    }
}

File: backend/dotnet/ticketflow/TicketFlow.TicketService/WebApi/Controllers/TicketsApiController.cs
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Microsoft.AspNetCore.Mvc;
using TicketFlow.Common.Readers;
using TicketFlow.TicketService.Domain.Entities;
using TicketFlow.TicketService.Domain.Models;
using TicketFlow.TicketService.Service;
using TicketFlow.TicketService.Service.Serializers;
using TicketFlow.TicketService.WebApi.ClientModels.Requests;

namespace TicketFlow.TicketService.WebApi.Controllers
{
    [ApiController]
    [Route("/tickets")]
    public class TicketsApiController : ControllerBase
    {
        private readonly ITicketService ticketService;
        private readonly IStringFromStreamReader stringFromStreamReader;
        private readonly ITicketSerializer ticketSerializer;

        public TicketsApiController(ITicketService ticketService, IStringFromStreamReader stringFromStreamReader, ITicketSerializer ticketSerializer)
        {
            this.ticketService = ticketService;
            this.stringFromStreamReader = stringFromStreamReader;
            this.ticketSerializer = ticketSerializer;
        }

        [HttpGet("/")]
        public IActionResult Index()
        {
            return new OkObjectResult("Hello from Ticket service API");
        }

        [HttpGet("by-movie/{movieId}")]
        public IReadOnlyCollection<TicketSerializationModel> GetByMovieId([FromRoute] int movieId)
        {
            IReadOnlyCollection<ITicket> tickets = ticketService.GetByMovieId(movieId);
            return tickets.Select(ticketSerializer.Serialize).ToArray();
        }

        [HttpPost("by-user")]
        public async Task<IReadOnlyCollection<TicketSerializationModel>> GetByUserEmail()
        {
            string userEmail = await stringFromStreamReader.ReadAsync(Request.Body, Encoding.UTF8);
            IReadOnlyCollection<ITicket> tickets = ticketService.GetByUserEmail(userEmail);
            return tickets.Select(ticketSerializer.Serialize).ToArray();
        }

        [HttpPost("")]
        public int Add([FromBody] AddTicketRequestModel addTicketRequestModel)
        {
            int createdEntityId = ticketService.Add(Convert(addTicketRequestModel));
            return createdEntityId;
        }

        [HttpPost("order")]
        public string Order([FromBody] OrderRequestModel orderRequestModel)
        {
            var orderModel = new OrderModel(orderRequestModel.TicketId, orderRequestModel.BuyerEmail);
            ticketService.Order(orderModel);

            return "Ordered successfully";
        }

        private static TicketCreationModel Convert(AddTicketRequestModel requestModel)
            => new TicketCreationModel(requestModel.MovieId, requestModel.Row, requestModel.Seat, requestModel.Price);
    }
}